base/sparse/higherorderfns.jl

```diff
@@ -913,33 +916,30 @@ _containertype{T<:Diagonal}(::Type{T}) = StructuredArray
 _containertype{T<:Bidiagonal}(::Type{T}) = StructuredArray
 _containertype{T<:Tridiagonal}(::Type{T}) = StructuredArray
 _containertype{T<:SymTridiagonal}(::Type{T}) = StructuredArray
-promote_containertype(::Type{StructuredArray}, ::Type{StructuredArray}) = StructuredArray
 # combinations involving sparse arrays continue in the structured array funnel
-promote_containertype(::Type{StructuredArray}, ::Type{AbstractSparseArray}) = StructuredArray
-promote_containertype(::Type{AbstractSparseArray}, ::Type{StructuredArray}) = StructuredArray
+promote_containertype(::Type{StructuredArray}, ::Type{AbstractSparseArray}) = AbstractSparseArray
+promote_containertype(::Type{AbstractSparseArray}, ::Type{StructuredArray}) = AbstractSparseArray
+promote_containertype(::Type{StructuredArray}, ::Type{VecOrMat}) = AbstractSparseArray
+promote_containertype(::Type{VecOrMat}, ::Type{StructuredArray}) = AbstractSparseArray
 # combinations involving scalars continue in the structured array funnel
-promote_containertype(::Type{StructuredArray}, ::Type{Any}) = StructuredArray
-promote_containertype(::Type{Any}, ::Type{StructuredArray}) = StructuredArray
+promote_containertype(::Type{StructuredArray}, ::ScalarType) = AbstractSparseArray
+promote_containertype(::ScalarType, ::Type{StructuredArray}) = AbstractSparseArray
 # combinations involving arrays divert to the generic array code
 promote_containertype(::Type{StructuredArray}, ::Type{Array}) = Array
 promote_containertype(::Type{Array}, ::Type{StructuredArray}) = Array
 # combinations involving tuples divert to the generic array code
 promote_containertype(::Type{StructuredArray}, ::Type{Tuple}) = Array
 promote_containertype(::Type{Tuple}, ::Type{StructuredArray}) = Array
 
-# for combinations involving sparse/structured arrays and scalars only,
-# promote all structured arguments to sparse and then rebroadcast
-@inline broadcast_c{N}(f, ::Type{StructuredArray}, As::Vararg{Any,N}) =
-    broadcast(f, map(_sparsifystructured, As)...)
-@inline broadcast_c!{N}(f, ::Type{AbstractSparseArray}, ::Type{StructuredArray}, C, B, As::Vararg{Any,N}) =
-    broadcast!(f, C, _sparsifystructured(B), map(_sparsifystructured, As)...)
-@inline broadcast_c!{N}(f, CT::Type, ::Type{StructuredArray}, C, B, As::Vararg{Any,N}) =
-    broadcast_c!(f, CT, Array, C, B, As...)
 @inline _sparsifystructured(S::SymTridiagonal) = SparseMatrixCSC(S)
 @inline _sparsifystructured(T::Tridiagonal) = SparseMatrixCSC(T)
 @inline _sparsifystructured(B::Bidiagonal) = SparseMatrixCSC(B)
 @inline _sparsifystructured(D::Diagonal) = SparseMatrixCSC(D)
-@inline _sparsifystructured(A::AbstractSparseArray) = A
+@inline _sparsifystructured(A::AbstractMatrix) = SparseMatrixCSC(A)
+@inline _sparsifystructured(A::AbstractVector) = SparseVector(A)
+@inline _sparsifystructured(A::AbstractSparseMatrix) = SparseMatrixCSC(A)
+@inline _sparsifystructured(A::AbstractSparseVector) = SparseVector(A)
+@inline _sparsifystructured(S::SparseVecOrMat) = S
```

Comment:
Good call. This combination of definitions (`_sparsifystructured` for `AbstractSparseMatrix`, `AbstractSparseVector`, and `SparseVecOrMat`) is safer than the definition that combination supplants (`_sparsifystructured` for `AbstractSparseArray`). We should make certain this change goes in one way or another. Thanks!